llvm/lib/ObjectYAML/GOFFEmitter.cpp

```diff
@@ -308,11 +308,14 @@ bool GOFFState::writeObject() {
   if (HasError)
     return false;
   // Iterate over all records.
-  for (const std::unique_ptr<llvm::GOFFYAML::RecordBase> &Rec : Doc.Records)
+  unsigned RecordNum = 0;
+  for (const std::unique_ptr<llvm::GOFFYAML::RecordBase> &Rec : Doc.Records) {
     if (const auto *Sym = dyn_cast<GOFFYAML::Symbol>(Rec.get()))
       writeSymbol(*Sym);
     else
-      reportError("unknown record type");
+      reportError("unknown record type on record index" + Twine(RecordNum));
```

Comment:
```suggestion
      reportError("record with index " + Twine(RecordNum) + " has unknown record type " + Twine(Rec.Type));
```
This is clearer, I think. `Rec.Type` may not be the right thing - I just used this as a placeholder so that I didn't have to look up the actual member that is used to determine what kind of Record it is.

You should also have a test case for this error.

Reply:
That is running in the wrong direction. The `if` is over the discriminator used for the inheritance hierarchy. There is no user input involved. The error message made no sense in the first place. The whole loop should be:

```
  for (const auto &Rec: Doc.Records) {
    GOFFYAML::RecordBase *RecBase = Rec.get();
    switch (RecBase->getKind()) {
      case GOFFYAML::RecordBase::RBK_Symbol:
        writeSymbol(*static_cast<GOFFYAML::Symbol *>(RecBase));
    }
  }
```

There is no `default` case because the `switch` covers all cases. I put this into the updated PR.